### PR TITLE
add spi trait

### PR DESCRIPTION
**Pull request by xoviat**

### CODE REVIEW

**Comment by xoviat** [comment]
what are the differences between this and [this](https://github.com/Disasm/async-embedded-traits/blob/master/src/spi.rs)? Seems this has more, but doesn't quite work. Might be better to just copy the async-embedded-traits trait for now...

**Comment by eupn** [comment]
It isn’t necessary to copy, we can reexport and implement async-embedded-traits here but also propose some changes to that crate regarding `Pin<&mut Self>` requirement.

**Comment by xoviat** [comment]
@Dirbaio is it fine to depend on async-embedded-traits?

**Comment by Dirbaio** [comment]
> @Dirbaio is it fine to depend on async-embedded-traits?

I'd prefer embassy-traits to stay standalone, for now at least. We have made quite a bit of changes to the traits, and we will very likely do more. Having everything in one repo means it's super easy to do a single PR updating the trait and all the impls at the same time. This helps a lot to catch badly designed traits faster: you can immediately find out a trait change is bad if you have issues updating an impl.

File: embassy-traits/src/i2c.rs [hunk 1]
```diff
@@ -92,7 +92,7 @@ impl AddressMode for SevenBitAddress {}
 impl AddressMode for TenBitAddress {}
 
 /// Blocking read
-pub trait Read<A: AddressMode = SevenBitAddress> {
+pub trait WriteRead<A: AddressMode = SevenBitAddress> {
```
Comment (Dirbaio) [inline embassy-traits/src/i2c.rs:95]: Can you remove unrelated changes?

**Comment by Dirbaio** [comment]
I think porting the nonblocking trait from embedded-hal is not the right choice:

- The "each `try_read` call must be preceded by a `try_send` call" restriction is *extremely* error-prone.
- Since each call only sends/receives one word, it's impossible for impls to effectively use DMA.

Maybe something like this could work:

```rust
fn read(mut self, read: &mut [W])
fn write(&mut self, write: &[W])
fn read_write(&mut self, read: &[W], write: &mut [W])
```

---

What's the reason for including Mode, Polarity, Phase? We can use them from embedded-hal, and they aren't used by the trait api itself.

**Comment by Dirbaio** [comment]
Looks great. Can you update the trait docs, they're now outdated?
